include/common/firmware/i2c.h

```diff
@@ -0,0 +1,29 @@
+#pragma once
```

Comment:
The more I read this the more I think the common code should just be hands off on specific peripheral initialization. This uses the same weird static-callback style that compilers are really bad at emitting diagnostics for as the hal, it makes you include the hal anyway if you include this file... I honestly think that if you had the c++ stuff exactly the same but just removed i2c.h and made the i2c_comms firmware object (this doesn't have to go in the concept or anything) need
- a fully constructed handle, by-reference (since i'm pretty sure it has internal pointers)
- and an associated address and address mode, which rather than being `#define`s were enum or enum class elements
then I think things would be better.

The other benefit to that is that it brings more of the static free-store initialization closer to itself. I think that sort of thing really should be kept in the end executables, those are the only places that should ship around references to static free store objects.

Reply:
In fact the other benefit of this is that you could even like forward declare your config element struct and then i2c_comms.hpp wouldn't have to include the hal at all, that'd be cool